stl/inc/ranges

```diff
@@ -6352,6 +6352,140 @@ namespace ranges {
+        } else if constexpr (_Ref_converts && constructible_from<_Container, from_range_t, _Rng, _Types...>) {
+            return _Container(from_range, _STD forward<_Rng>(_Range), _STD forward<_Types>(_Args)...);
+        } else if constexpr (_Ref_converts && _Common_constructible<_Rng, _Container, _Types...>) {
+            return _Container(_RANGES begin(_Range), _RANGES end(_Range), _STD forward<_Types...>(_Args)...);
+        } else if constexpr (_Ref_converts && _Constructible_and_insertable<_Rng, _Container, _Types...>) {
+            _Container _Cont(_STD forward<_Types>(_Args)...);
+            if constexpr (_Sized_and_reservable<_Rng, _Container>) {
+                _Cont.reserve(_RANGES size(_Range));
+            }
+            _RANGES copy(_Range, _Container_inserter<range_reference_t<_Rng>>(_Cont));
+            return _Cont;
+        } else if constexpr (input_range<range_reference_t<_Rng>>) {
+            const auto _Xform = [](auto&& _Elem) {
+                return _RANGES to<range_value_t<_Container>>(_STD forward<decltype(_Elem)>(_Elem));
+            };
+            return _RANGES to<_Container>(views::transform(_Range, _Xform), _STD forward<_Types>(_Args)...);
+        } else {
+            static_assert(_Always_false<_Container>, "the program is ill-formed per N4910 [range.utility.conv.to]/1.3");
+        }
+    }
+
+    template <class _Container>
+    struct _To_class_fn {
+        _STL_INTERNAL_STATIC_ASSERT(!view<_Container>);
+
+        template <input_range _Rng, class... _Types>
+        _NODISCARD constexpr auto operator()(_Rng&& _Range, _Types&&... _Args) const requires requires {
+            _RANGES to<_Container>(_STD forward<_Rng>(_Range), _STD forward<_Types>(_Args)...);
+        }
+        { return _RANGES to<_Container>(_STD forward<_Rng>(_Range), _STD forward<_Types>(_Args)...); }
+    };
+
+    // clang-format off
+    template <class _Container, class... _Types>
+        requires (!view<_Container>)
+    _NODISCARD constexpr auto to(_Types&&... _Args) {
+        return _Range_closure<_To_class_fn<_Container>, decay_t<_Types>...>{_STD forward<_Types>(_Args)...};
+    }
+    // clang-format on
+
+    template <input_range _Rng>
+    struct _Phony_input_iterator {
+        using iterator_category = input_iterator_tag;
+        using value_type        = range_value_t<_Rng>;
+        using difference_type   = ptrdiff_t;
+        using pointer           = add_pointer_t<range_reference_t<_Rng>>;
+        using reference         = range_reference_t<_Rng>;
+
+        reference operator*() const;
+        pointer operator->() const;
+
+        _Phony_input_iterator& operator++();
+        _Phony_input_iterator operator++(int);
+
+        bool operator==(const _Phony_input_iterator&) const;
+    };
+
+    template <template <class...> class _Container, class _Rng, class... _Types>
+    auto _To_helper(int) -> decltype(_Container(_STD declval<_Rng>(), _STD declval<_Types>()...)); // not defined

```

Comment:
and now with named concepts: https://gcc.godbolt.org/z/KraW45TEK

and more sfinae: https://gcc.godbolt.org/z/eEEY9W4fe

Reply:
Did some compile benchmarks, and I couldn't find a difference between the implementations outside noise (and waiting for cl to start up...)

Reply:
I believe that this
```cpp
namespace CtorType {
    struct Basic;
    struct FromRange;
    struct FromIters;

    template <class C>
    struct CtorType_impl : std::false_type {};
    template <>
    struct CtorType_impl<Basic> : std::true_type {};
    template <>
    struct CtorType_impl<FromRange> : std::true_type {};
    template <>
    struct CtorType_impl<FromIters> : std::true_type {};

    template <class C>
    concept t = CtorType_impl<C>::value;
}
```

should rather be something like
```cpp
enum class CtorType {
    Basic,
    FromRange,
    FromIter,
}

template<CtorType>
inline constexpr bool someFittingName_v = false;

template<>
inline constexpr bool someFittingName_v<CtorType::Basic> = true;
template<>
inline constexpr bool someFittingName_v<CtorType::FromRange> = true;
template<>
inline constexpr bool someFittingName_v<CtorType::FromIter> = true;

template <CtorType C>
concept t = someFittingName_v<C>;
```

Reply:
I'm trying to avoid the full CPO style since we don't care about `noexcept` for `ranges::to`. This is admittedly lazy, in that all three helper overloads are specialized and perform TAD for overload resolution even though we might only need one or two.

We could go full-on "only template aliases" mode (https://gcc.godbolt.org/z/r4MMjb78T), but I feel that's too much code for something that's likely to have hundreds instead of thousands of calls. I steal your approach, but transport the types through pointers to avoid instantiating `type_identity` specializations (https://github.com/microsoft/STL/pull/2870).

Reply:
FYI, Measurements show effectively no difference between your "encapsulate the overload set in a constexpr function" approach and my "everything is a template alias" approach. I'm happy =)